design/downstream-crl-design.md

```diff
@@ -0,0 +1,128 @@
+# Downstream Certificate Revocation List Proposal
+
+Status: Draft
+
+## Abstract
+This proposal covers the implementation of CRLs (Certificate Revocation List) in Contours DownstreamValidation, using the `crl` field in Envoy's [extensions.transport_sockets.tls.v3.CertificateValidationContext](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-certificatevalidationcontext) field.
+
+## Background
+
+Envoy supports specifying a PEM encoded CRL for use when validating a peer certificate.
+Currently Contour allows specifying a CA Certificate bundle that is used to validate a client.
+This is done using Envoy's Common TLS Context inside of a Downstream TLS context.
+Envoy's Common TLS Context also supports a CRL data source, which is a set of CRL's that are used when validating a clients certificate.
+This CRL data source can be exposed the same way that we expose the CA Certificate data source, through Contour's `DownstreamValidation` structure.
+This data source will give admins the ability to have finer grained control over their ingress, and enable CRLs for HTTPProxies that use client cert authentication
+
+## Goals
+- Allow a Kubernetes secret containing a PEM encoded CRL to be used when validating client connections
+- Perform simple validation of the K8s secret
+  - Validate that it is indeed a PEM CRL object
+
+## Non Goals
+- Upstream Validation
+- Support for `only_verify_leaf_cert_crl`
+  - Not in a released version of envoy yet, still in development
+
+
+## High-Level Design
+The design would for the most part mirror the implementation of `DownstreamValidation.CACertificate` and `DownstreamValidation.SkipClientCertValidation`.
+
+Sample YAML
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: client-validation-example
+  namespace: default
+spec:
+  virtualhost:
+    fqdn: example.com
+    tls:
+      secretName: server-credentials
+      clientValidation:
+        caSecret: ca-cert-for-client-validation
+        crlSecret: crl-for-client-validation
+  routes:
+    - services:
+        - name: service
+          port: 80
+```
+
+## Detailed Design
+
+### CRL Secret
+
+The same approach shall be followed for configuring revocation lists as is used currently to store the CA certificates for client validation
+
+The CRL is stored in an opaque Kubernetes secret.
+The secret will be stored in the same namespace as the corresponding `HTTPProxy` object.
+The secret object shall contain entry named `crl.pem`.
```

Comment:
I think @sunjayBhatia mentioned this in one of his comments, but would storing this in the CA cert secret make any sense, or do you prefer to keep it separate? Keeping it separate definitely has the advantage of giving you more space to work with, but just wondering if you'd considered the alternative.

Reply:
I thought about it, but I think its an added bit of confusion. Having it be a separate secret is clear from a configuration point of view. I guess technically the user could put the same secret for both?

Reply:
Fair point, as long as the validation allows it which it currently does. In any case I am fine with the separate secret.